packages/yswaps/contracts/TradeFactory/TradeFactoryExecutor.sol

```diff
@@ -69,77 +74,84 @@ abstract contract TradeFactoryExecutor is ITradeFactoryExecutor, TradeFactoryPos
   }
 
   // Execute via sync swapper
-  function execute(
-    address _tokenIn,
-    address _tokenOut,
-    uint256 _amountIn,
-    uint256 _maxSlippage,
-    bytes calldata _data
-  ) external override onlyRole(STRATEGY) returns (uint256 _receivedAmount) {
+  function execute(SyncTradeExecutionDetails calldata _tradeExecutionDetails, bytes calldata _data)
+    external
+    override
+    onlyRole(STRATEGY)
+    returns (uint256 _receivedAmount)
+  {
     address _swapper = strategySyncSwapper[msg.sender];
-    if (_tokenIn == address(0) || _tokenOut == address(0)) revert CommonErrors.ZeroAddress();
-    if (_amountIn == 0) revert CommonErrors.ZeroAmount();
-    if (_maxSlippage == 0) revert CommonErrors.ZeroSlippage();
-    IERC20(_tokenIn).safeTransferFrom(msg.sender, _swapper, _amountIn);
-    uint256 _preSwapBalanceOut = IERC20(_tokenOut).balanceOf(msg.sender);
-    ISyncSwapper(_swapper).swap(msg.sender, _tokenIn, _tokenOut, _amountIn, _maxSlippage, _data);
-    _receivedAmount = IERC20(_tokenOut).balanceOf(msg.sender) - _preSwapBalanceOut;
-    emit SyncTradeExecuted(msg.sender, _swapper, _tokenIn, _tokenOut, _amountIn, _maxSlippage, _receivedAmount);
+    if (_tradeExecutionDetails._tokenIn == address(0) || _tradeExecutionDetails._tokenOut == address(0)) revert CommonErrors.ZeroAddress();
+    if (_tradeExecutionDetails._amountIn == 0) revert CommonErrors.ZeroAmount();
+    if (_tradeExecutionDetails._maxSlippage == 0) revert CommonErrors.ZeroSlippage();
+    IERC20(_tradeExecutionDetails._tokenIn).safeTransferFrom(msg.sender, _swapper, _tradeExecutionDetails._amountIn);
+    uint256 _preSwapBalanceOut = IERC20(_tradeExecutionDetails._tokenOut).balanceOf(msg.sender);
+    ISyncSwapper(_swapper).swap(
+      msg.sender,
+      _tradeExecutionDetails._tokenIn,
+      _tradeExecutionDetails._tokenOut,
+      _tradeExecutionDetails._amountIn,
+      _tradeExecutionDetails._maxSlippage,
+      _data
+    );
+    _receivedAmount = IERC20(_tradeExecutionDetails._tokenOut).balanceOf(msg.sender) - _preSwapBalanceOut;
+    emit SyncTradeExecuted(_tradeExecutionDetails, msg.sender, _receivedAmount, _swapper);
   }
 
   // Execute via async swapper
   function execute(
-    address _strategy,
-    address _tokenIn,
-    address _tokenOut,
-    uint256 _amountIn,
+    AsyncTradeExecutionDetails calldata _tradeExecutionDetails,
     address _swapper,
-    uint256 _minAmountOut,
     bytes calldata _data
   ) external override onlyMechanic returns (uint256 _receivedAmount) {
-    if (!_tokensOutByStrategyAndTokenIn[_strategy][_tokenIn].contains(_tokenOut)) revert InvalidTrade();
+    if (
+      !_tokensOutByStrategyAndTokenIn[_tradeExecutionDetails._strategy][_tradeExecutionDetails._tokenIn].contains(
+        _tradeExecutionDetails._tokenOut
+      )
+    ) revert InvalidTrade();
     if (!_swappers.contains(_swapper)) revert InvalidSwapper();
-    if (_amountIn == 0) _amountIn = IERC20(_tokenIn).balanceOf(_strategy);
```

Comment:
I know you probably loved this line, the issue is: When I converted input to struct, I can no longer re-write on the input param `amount`. I can create a new `uint256` and do this if, probably a lil bit more costly. Tell me wdyt.

Reply:
yes, please let's keep the same functionality